modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java

```diff
@@ -166,45 +171,110 @@ private void ensureOpen() {
     }
 
     private void reopenStreamOrFail(IOException e) throws IOException {
-        if (purpose == OperationPurpose.REPOSITORY_ANALYSIS) {
-            logger.warn(() -> format("""
-                failed reading [%s/%s] at offset [%s]""", blobStore.bucket(), blobKey, start + currentOffset), e);
-            throw e;
+        maybeRecordAndLogForRetry("reading", e);
+        maybeAbort(currentStream);
+        IOUtils.closeWhileHandlingException(currentStream);
+
+        maybeDelaySafely();
+        openStreamWithRetry();
+    }
+
+    private <T extends Exception> void maybeRecordAndLogForRetry(String action, T e) throws T {
+        if (shouldRetry() == false) {
+            final var finalException = addSuppressedExceptions(e);
+            logForRetry(Level.WARN, action, finalException);
+            throw finalException;
         }
 
-        final int maxAttempts = blobStore.getMaxRetries() + 1;
+        attempt += 1;
+        // Log at info level every ~5 minutes
+        logForRetry(attempt % 30 == 0 ? Level.INFO : Level.DEBUG, action, e);
```

Comment:
Can we ensure the first retry is always output on INFO? Perhaps something like:

`Integer.bitCount(attempt) == 1`?

which would log exponentially less.

Reply:
I updated it to log the very first retry with
```
logForRetry((attempt == 2 || attempt % 30 == 0) ? Level.INFO : Level.DEBUG, action, e);
```
Note the `attempt` starts with `1` as it is a count for all attempts including the original one and subsequent retries so that the first retry is the 2nd attempt.

I prefer to keep logging about every 5 minutes since the behaviour would be more predicatble in cases where we experience extended period of s3/network outages.

Reply:
I am ok with that, but notice that if there is an s3 outage, we could have many threads doing this per node, perhaps 50 or so currently, perhaps more in the future for large nodes. Still just a message every 6s (per node), so feels ok'ish. We can refine later too ofc.

Reply:
Potential excessive logging from multiple threads is a good point. I can adjust this again in the follow-up where I will also add log messages when the retry eventually succeeds (see also https://github.com/elastic/elasticsearch/pull/103300#issuecomment-1882130933).